starky/src/lookup.rs

```diff
@@ -761,79 +761,35 @@ pub(crate) fn get_helper_cols<F: Field>(
         .div_ceil(constraint_degree.checked_sub(1).unwrap_or(1));
 
-    let mut helper_columns = Vec::with_capacity(num_helper_columns);
-
-    for mut cols_filts in &columns_filters
-        .iter()
-        .chunks(constraint_degree.checked_sub(1).unwrap_or(1))
-    {
-        let (first_col, first_filter) = cols_filts.next().unwrap();
-
-        let mut filter_col = Vec::with_capacity(degree);
-        let first_combined = (0..degree)
-            .map(|d| {
-                let f = {
-                    let f = first_filter.eval_table(trace, d);
-                    filter_col.push(f);
-                    f
-                };
-                if f.is_one() {
-                    let evals = first_col
-                        .iter()
-                        .map(|c| c.eval_table(trace, d))
+    let chunks = columns_filters.chunks(constraint_degree.checked_sub(1).unwrap_or(1));
+    let helper_columns: Vec<_> = chunks
+        .filter_map(|cols_filts| {
+            cols_filts
+                .iter()
+                .map(|(col, filter)| {
+                    let combined = (0..degree)
+                        .map(|d| {
+                            let evals = col
+                                .iter()
+                                .map(|c| c.eval_table(trace, d))
+                                .collect::<Vec<F>>();
+                            challenge.combine(&evals)
+                        })
                         .collect::<Vec<F>>();
-                    challenge.combine(evals.iter())
-                } else {
-                    assert_eq!(f, F::ZERO, "Non-binary filter?");
-                    // Dummy value. Cannot be zero since it will be batch-inverted.
-                    F::ONE
-                }
-            })
-            .collect::<Vec<F>>();
 
-        let mut acc = F::batch_multiplicative_inverse(&first_combined);
-        for d in 0..degree {
-            if filter_col[d].is_zero() {
-                acc[d] = F::ZERO;
-            }
-        }
-
-        for (col, filt) in cols_filts {
-            let mut filter_col = Vec::with_capacity(degree);
-            let mut combined = (0..degree)
-                .map(|d| {
-                    let f = {
-                        let f = filt.eval_table(trace, d);
-                        filter_col.push(f);
-                        f
-                    };
-                    if f.is_one() {
-                        let evals = col
-                            .iter()
-                            .map(|c| c.eval_table(trace, d))
-                            .collect::<Vec<F>>();
-                        challenge.combine(evals.iter())
-                    } else {
-                        assert_eq!(f, F::ZERO, "Non-binary filter?");
-                        // Dummy value. Cannot be zero since it will be batch-inverted.
-                        F::ONE
-                    }
+                    let mut combined = F::batch_multiplicative_inverse(&combined);
```

Comment:
If I remember correctly, the reason why the previous code did not use `batch_multiplicative_inverse` (and was therefore more complicated). was because the filter column is often (not always!) sparse, so we wanted to avoid inverting a lot of values that would be later zeroed out. However, I am not sure how much of an impact it would actually have, and this code does look much nicer!

Reply:
Yes, though the code in main already uses batch multiplicative inverse.